src/write/encoder_utf8.rs

```diff
@@ -0,0 +1,187 @@
+use std::fmt;
+use std::io;
+use std::str::from_utf8;
+
+use crate::engine::Engine;
+use crate::write::EncoderWriter;
+
+/// A [`io::Write`] wrapper for types implementing [`fmt::Write`]
+///
+/// It needn't be used directly, but as a parameter for [`EncoderWriter`].
+///
+/// # Examples
+///
+/// Write base64 into a new [`String`]:
+///
+/// ```
+/// use std::io::Write;
+/// use base64::engine::general_purpose;
+///
+/// let mut enc = base64::write::EncoderWriter::string(&general_purpose::STANDARD);
+///
+/// enc.write_all(b"asdf").unwrap();
+///
+/// // get the resulting String
+/// let b64_string = enc.formatter();
+///
+/// assert_eq!("YXNkZg==", &b64_string);
+/// ```
+///
+/// Or, append to an existing [`String`], which implements [`fmt::Write`]:
+///
+/// ```
+/// use std::io::Write;
+/// use base64::engine::general_purpose;
+///
+/// let mut buf = String::from("base64: ");
+///
+/// let mut enc = base64::write::EncoderWriter::utf8(
+///     &mut buf,
+///     &general_purpose::STANDARD);
+///
+/// enc.write_all(b"asdf").unwrap();
+///
+/// // release the &mut reference on buf
+/// let _ = enc.formatter();
+///
+/// assert_eq!("base64: YXNkZg==", &buf);
+/// ```
+///
+/// # Performance
+///
+/// Because it has to validate that the base64 is UTF-8, it is about 80% as fast as writing plain
+/// bytes to a `io::Write`.
+pub struct Utf8Compat<W: fmt::Write> {
+    inner: W,
+}
+
+impl<W: fmt::Write> Utf8Compat<W> {
+    /// Create wrapper implementing [`io::Write`] for [`fmt::Write`]
+    pub fn new(writer: W) -> Self {
+        Self { inner: writer }
+    }
+
+    /// Extract the underlying writer
+    pub fn writer(self) -> W {
+        self.inner
+    }
+}
+
+impl<W: fmt::Write> From<W> for Utf8Compat<W> {
+    fn from(value: W) -> Self {
+        Self::new(value)
+    }
+}
+
+impl<W: fmt::Write> io::Write for Utf8Compat<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.inner
+            .write_str(from_utf8(buf).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?)
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))
+            .map(|()| buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl<'e, E: Engine, W: fmt::Write> EncoderWriter<'e, E, Utf8Compat<W>> {
+    /// Create a [`EncoderWriter`] that will write [`&str`]
+    pub fn utf8(writer: W, engine: &'e E) -> Self {
+        Self::new(Utf8Compat::new(writer), engine)
+    }
+
+    /// Encode all remaining buffered data, including any trailing incomplete input triples and
+    /// associated padding.
+    ///
+    /// Returns the base64-encoded form of the accumulated written data.
+    pub fn formatter(mut self) -> W {
+        self.finish()
+            .expect("Writing to a consumer should never fail")
```

Comment:
This is what's not going to work for me. We should not assume that this will always succeed. Anyone is free to implement Write and take the contract at its word that write_str can fail.